File: apps/cli/src/serve.ts
/**
 * `refyard serve` / `refyard open` — run an assembled service in the foreground.
 *
 * The service itself is assembled by `@refyard/host-node`; what stays here is the part
 * that belongs to a command-line host: the port policy, the pairing control socket, the
 * terminal banner with the pairing URL, and the signal handlers.
 *
 * Foreground is the whole lifecycle model in this version: the process runs until
 * Ctrl+C, and Ctrl+C stops accepting new work before it exits. There is no daemon,
 * no service installer and no supervision — Xross is the future supervisor, and it
 * is not this release.
 */
import {
  createInterface,
  type Interface as ConsoleInterface,
} from "node:readline";
import {
  assembleService,
  defaultStateRoot,
  PortInUseError,
  startHttpHost,
  startPairingControl,
  type AssembleOptions,
  type HttpHost,
  type HttpHostOptions,
  type ServiceAssembly,
} from "@refyard/host-node";
import { API_MAJOR, CONTRACT_VERSION } from "@refyard/git-contract";
import { DEFAULT_PORT, DEFAULT_TICKET_TTL_SECONDS } from "./args.js";

import { isPairingCommand } from "./pairing-reprint.js";
import { reportedVersion } from "./version.js";
import { openInBrowser } from "./browser.js";


export interface RunServiceOptions extends AssembleOptions {
  readonly port: number;
  /**
   * Whether `port` was asked for by name.
   *
   * An explicit port is a request for *that* port — a bookmark, a tunnel, a script that
   * expects it — so a busy one is refused. The default port is a courtesy, and a busy one
   * means "take another free port and say which". Required rather than defaulted: a call
   * site that has not thought about which case it is in should not silently get one.
   */
  readonly portExplicit: boolean;
  readonly openBrowser: boolean;
  /** Pairing-ticket lifetime in seconds; 60 unless --ticket-ttl says otherwise. */
  readonly ticketTtlSeconds: number;
  /** Optional static root for an embedding host; the CLI itself leaves this absent. */
  readonly webRoot?: string | null;
  /** Exact browser origins allowed to call this API in hosted mode. */
  readonly allowedOrigins?: readonly string[];
  /** Origin of the separately deployed UI that receives the pairing URL. */
  readonly uiOrigin?: string;
  /** Browser-visible API origin, such as the HTTPS endpoint exposed by a tunnel. */
  readonly apiOrigin?: string;
  /** Hosted pairing secret, supplied by the environment and never placed in argv or logs. */
  readonly hostedPassword?: string;
  readonly allowRoot: boolean;
  /**
   * Machine output: one JSON object on stdout, pairing material on stderr only.
   *
   * A terminal is a private channel; a pipe may be read by anything, so the ticket
   * never appears on stdout in this mode.
   */
  readonly json?: boolean;
  /** Injected by tests so a run does not need a signal to stop. */
  readonly installSignalHandlers?: boolean;
  /** How long a stop waits for in-flight requests; the host's default is 5 s. */
  readonly shutdownGraceMs?: number;
}

export interface RunningService {
  readonly http: HttpHost;
  readonly assembly: ServiceAssembly;
  readonly url: string;
  /** The pairing URL carrying the one-time ticket. */
  readonly pairingUrl: string;
  close(): Promise<void>;
}

export async function runService(
  options: RunServiceOptions,
): Promise<RunningService> {
  if (
    process.getuid !== undefined &&
    process.getuid() === 0 &&
    !options.allowRoot
  ) {
    throw new Error(
      "refyard refuses to run as root: Git hooks and filters would run with root privileges. Pass --allow-root only if you understand that.",
    );
  }

  const assembly = await assembleService(options);
  const note = options.writeError ?? options.write;
  // Annotated so a mistyped field is a compile error: an unannotated object spread into
  // the call would silently drop whatever the host does not recognise.
  const hostOptions: Omit<HttpHostOptions, "port"> = {
    read: assembly.read,
    mutations: assembly.mutations,
    events: assembly.events,
    serviceInstanceId: assembly.serviceInstanceId,
    ticketTtlSeconds: options.ticketTtlSeconds,
    webRoot: options.webRoot ?? null,
    allowedOrigins: [
      ...(options.allowedOrigins ?? []),
      ...(options.uiOrigin === undefined ? [] : [options.uiOrigin]),
    ],
    ...(options.hostedPassword === undefined
      ? {}
      : { hostedPassword: options.hostedPassword }),
    repositoryManagement: assembly.repositoryManagement,
    ...(options.shutdownGraceMs === undefined
      ? {}
      : { shutdownGraceMs: options.shutdownGraceMs }),
    repositoryRootOf: (repositoryId: string) =>
      assembly.repositories.get(repositoryId)?.allowedRootId ?? null,
    provider: assembly.provider,
    grants: {
      allowedRootIds: [...assembly.allowedRootIds],
      repositoryIds: [...assembly.repositoryIds],
      scopes: [
        "repository:read",
        "repository:write",
        "repository:network",
        "workspace:manage",
        "provider:manage",
      ],
    },
    log: (line) => {
      // Service logs are diagnostics, not data: a machine reading stdout must not
      // find them interleaved with the JSON object.
      if (options.json === true) {
        note(`  ${line}`);
      } else {
        options.write(`  ${line}`);
      }
    },
  };

  let http: HttpHost;
  try {
    http = await startHttpHost({ ...hostOptions, port: options.port });
  } catch (error) {
    if (error instanceof PortInUseError && !options.portExplicit) {
      // The default port is a courtesy, not a promise: a second refyard, a dev server or a
      // tunnel may hold it. Taking a free port instead keeps the default usable, and the
      // note names both ports so nothing moves silently. An explicitly requested port is
      // not retried — see `RunServiceOptions.portExplicit`.
      http = await startHttpHost({ ...hostOptions, port: 0 });
      note(
        `note: port ${error.port} is in use; listening on port ${http.port} instead (pass --port to pin one)`,
      );
    } else {
      throw error;
    }
  }

  const origin = `http://127.0.0.1:${http.port}`;
  const pairingUrlFor = (): string => {
    const pairingOrigin = options.uiOrigin ?? origin;
    const issued = http.pairingUrl(pairingOrigin);
    if (options.uiOrigin === undefined) {
      return issued;
    }
    const url = new URL(issued);
    url.searchParams.set("api", options.apiOrigin ?? origin);
    return url.toString();
  };
  const pairingUrl = pairingUrlFor();
  const localUi = options.webRoot !== undefined && options.webRoot !== null;

  /**
   * The trusted local channel for more tickets: `refyard pair` connects to this
   * socket (same user only) and asks for a fresh pairing URL — the programmatic
   * twin of the `p` keystroke. Minting stays off HTTP on purpose: an endpoint
   * that hands out credentials would widen the surface the ticket protects.
   */
  const pairingControl = await startPairingControl({
    stateRoot: options.stateRootPath ?? defaultStateRoot(),
    instanceId: http.serviceInstanceId,
    url: origin,
    port: http.port,
    mintPairingUrl: pairingUrlFor,
    onMinted: () => {
      // The ticket itself goes to the requester; the service's own terminal
      // keeps a line so a human can see that one was handed out.
      note("a fresh pairing URL was issued over the local control socket (`refyard pair`)");
    },
  });

  /**
   * Two output modes, and the difference is who is reading.
   *
   * A terminal gets the banner, including the pairing URL — the user's own screen is
   * the private channel that a web page cannot read. A machine (`--json`) gets one
   * JSON object on stdout with no pairing material in it at all, and the pairing URL
   * goes to stderr: a supervisor can capture it, a log can strip it, and a program
   * parsing stdout can never mistake a ticket for data.
   */
  if (options.json === true) {
    options.write(
      JSON.stringify({
        serviceInstanceId: http.serviceInstanceId,
        port: http.port,
        url: origin,
        apiMajor: API_MAJOR,
        contractVersion: CONTRACT_VERSION,
        repositoryId: assembly.repositoryId,
        repositoryIds: [...assembly.repositoryIds],
        allowedRootIds: [...assembly.allowedRootIds],
        ui: localUi ? origin : null,
        apiOnly: !localUi,
      }),
    );
    note(`pairing URL (single use): ${pairingUrl}`);
    if (options.openBrowser && (localUi || options.uiOrigin !== undefined)) {
      const opened = await openInBrowser(pairingUrl);
      if (!opened.ok) {
        note(`could not open a browser: ${opened.reason}`);
      }
    } else if (options.openBrowser) {
      note(
        "not opening a browser: this CLI serves the API only; configure the deployed UI origin with --ui-origin",
      );
    }
    let jsonClosed = false;
    const closeJson = async (): Promise<void> => {
      if (jsonClosed) {
        return;
      }
      jsonClosed = true;
      await http.close();
      await pairingControl.close();
    };
    if (options.installSignalHandlers !== false) {
      const onJsonSignal = (): void => {
        note("stopping: no new requests will be accepted");
        void closeJson().then(() => {
          process.exitCode = 0;
        });
      };
      process.on("SIGINT", onJsonSignal);
      process.on("SIGTERM", onJsonSignal);
    }
    return { http, assembly, url: origin, pairingUrl, close: closeJson };
  }

  options.write(`refyard ${await reportedVersion()} (api ${API_MAJOR})`);
  options.write(`  repositories: ${assembly.repositoryPaths.join(", ")}`);
  options.write(`  git:        ${assembly.gitVersion}`);
  options.write(
    `  port:       ${http.port}${options.port === 0 ? " (chosen by the OS)" : ""}`,
  );
  options.write(
    `  ticket ttl: ${options.ticketTtlSeconds}s${options.ticketTtlSeconds === DEFAULT_TICKET_TTL_SECONDS ? "" : " (--ticket-ttl)"}`,
  );
  options.write(
    localUi
      ? `  ui:         ${origin} (bundled local workbench)`
      : "  ui:         API-only (use a hosted UI or integration client)",
  );
  options.write("");
  options.write(`  open this URL in your browser to pair this session:`);
  options.write(`    ${pairingUrl}`);
  options.write("");
  options.write(
    `  ready ${JSON.stringify({ serviceInstanceId: http.serviceInstanceId, port: http.port, url: origin })}`,
  );
  options.write(
    `  press p + Enter to print another pairing URL (each is single use)`,
  );
  options.write(
    `  or run \`refyard pair\` in another terminal for the same thing`,
  );
  options.write(`  press Ctrl+C to stop`);

  if (options.openBrowser && (localUi || options.uiOrigin !== undefined)) {
    const opened = await openInBrowser(pairingUrl);
    if (!opened.ok) {
      // The URL is already printed, so a missing browser is a note, not a failure.
      options.write(
        `  note: could not open a browser (${opened.reason}); use the URL above`,
      );
    }
  } else if (options.openBrowser) {
    options.write(
      "  note: not opening a browser: this CLI serves the API only; configure --ui-origin for the deployed UI",
    );
  }

  let closed = false;
  // Created once the listener is up; `close` shuts it down so the terminal
  // interface never holds the process open after a stop.
  let consoleInterface: ConsoleInterface | null = null;
  const close = async (): Promise<void> => {
    if (closed) {
      return;
    }
    closed = true;
    // Stop accepting new requests first, then drop sessions: an in-flight read gets
    // to finish, and nothing new is paired after the user asked to stop.
    await http.close();
    await pairingControl.close();
    consoleInterface?.close();
  };

  if (options.installSignalHandlers !== false) {
    const onSignal = (): void => {
      options.write("");
      options.write("stopping: no new requests will be accepted");
      void close().then(() => {
        process.exitCode = 0;
      });
    };
    process.on("SIGINT", onSignal);
    process.on("SIGTERM", onSignal);

    // A second browser needs a second ticket, and minting must stay out of the
    // authenticated API: the user's terminal is the channel that cannot be spoofed by a
    // web page. The interface is closed with the service so it never holds the process
    // open after a stop.
    const readline = createInterface({ input: process.stdin });
    consoleInterface = readline;
    readline.on("line", (line) => {
      if (isPairingCommand(line)) {
        options.write("");
        options.write("  a fresh pairing URL (single use):");
        options.write(`    ${pairingUrlFor()}`);
      }
    });
  }

  return {
    http,
    assembly,
    url: origin,
    pairingUrl,
    close,
  };
}

export { DEFAULT_PORT };

File: crates/refyard-cli/src/serve.rs
//! `serve` and `open`: the native process that exposes the service over loopback HTTP.
//!
//! This is the composition root for the native command line, and the same one the
//! desktop window uses: the Git the machine has, the private state directory for the
//! journal, the SSH configuration the environment names. On top of that it adds the two
//! decisions a serving process makes before the first request — which repositories were
//! approved, and therefore what a paired session may reach.
//!
//! Foreground is the whole lifecycle: the process runs until Ctrl+C, and a stop refuses
//! new work before it ends. There is no daemon and no installer.

use std::path::PathBuf;
use std::sync::Arc;

use refyard_host::providers::local::LocalGit;
use refyard_host::service::{ApplicationService, ApplicationServiceConfig};
use refyard_host::state_root::default_state_root;
use refyard_http::auth::Grants;

/// Where a serving run writes. Callbacks rather than streams, because the readiness
/// line has to reach the supervisor *while the process runs* — a buffered answer that
/// arrives at exit is indistinguishable from never answering.
#[derive(Clone)]
pub struct ServeOutput {
    pub stdout: Arc<dyn Fn(&str) + Send + Sync>,
    pub stderr: Arc<dyn Fn(&str) + Send + Sync>,
}

/// One serving run's options, already parsed.
pub struct ServeOptions {
    pub port: u16,
    /// Whether `port` was named on the command line: an explicit port is a request for
    /// *that* port, so a busy one is refused rather than silently moved.
    pub port_explicit: bool,
    pub ticket_ttl_seconds: u64,
    /// Machine output: one JSON object on stdout, the pairing URL on stderr.
    pub json: bool,
    /// Bind the pairing ticket to the empty origin: a supervisor spends it over a plain
    /// fetch with no Origin header, and no browser can spend it at all.
    pub machine: bool,
    /// Open the pairing URL in this machine's browser once the listener is up.
    pub open_browser: bool,
    /// The built workbench to serve from `/`; `None` makes this API-only.
    pub web_root: Option<PathBuf>,
    /// Repositories approved before the first request.
    pub paths: Vec<String>,
    pub output: ServeOutput,
}

/// Serves until the process is asked to stop, answering the exit status.
pub async fn run(options: ServeOptions) -> i32 {
    let output = options.output.clone();
    let stdout = move |line: &str| (output.stdout)(line);
    let stderr = move |line: &str| (output.stderr)(line);

    let git = match LocalGit::discover() {
        Ok(git) => git,
        Err(message) => {
            stderr(&format!("refyard: {message}"));
            return 2;
        }
    };
    let environment: Vec<(String, String)> = std::env::vars().collect();
    let ssh_config = ssh_config_from(&environment);
    let state_root = default_state_root(&environment, std::env::consts::OS);
    let home = refyard_host::reads::filesystem::home_from(git.environment())
        .unwrap_or_else(|| PathBuf::from("/"));
    let started = refyard_host::clock::now_millis();
    let service_builder = ApplicationService::new(ApplicationServiceConfig {
        git,
        service_instance_id: format!("srvc_{:x}{:x}", std::process::id(), started),
        target_id: "tgt_local".to_owned(),
        target_generation: format!("gen_{started}"),
        home,
    })
    .with_state_root(state_root);
    let service_builder = match service_builder {
        Ok(service) => service.with_ssh_config_opt(ssh_config),
        Err(problem) => {
            stderr(&format!(
                "refyard: the private state could not be opened: {problem}"
            ));
            return 2;
        }
    };
    let service = Arc::new(service_builder.with_writes());

    // The approved repositories are the grant: what was named on the command line is
    // what a paired session may reach. Repositories registered later through the API
    // extend the registering session's own grants, never everyone's.
    let mut allowed_root_ids = Vec::new();
    let mut repository_ids = Vec::new();
    for path in &options.paths {
        // The service answers with the whole registry, and a person's spelling of the
        // path is not the service's display form of it (a `/tmp` link is `/private/tmp`
        // behind the name), so the newly approved repository is found by difference, not
        // by comparing spellings.
        let before: std::collections::BTreeSet<String> = service
            .repositories()
            .await
            .repositories
            .into_iter()
            .map(|record| record.repository_id)
            .collect();
        match service.register_repository_on(path, None).await {
            Ok(answer) => {
                for record in answer.repositories {
                    if !before.contains(&record.repository_id)
                        && !repository_ids.contains(&record.repository_id)
                    {
                        repository_ids.push(record.repository_id.clone());
                    }
                    if !before.contains(&record.repository_id)
                        && !allowed_root_ids.contains(&record.allowed_root_id)
                    {
                        allowed_root_ids.push(record.allowed_root_id.clone());
                    }
                }
            }
            Err(problem) => {
                stderr(&format!(
                    "refyard: {path} could not be approved: {}",
                    problem.message
                ));
                return 2;
            }
        }
    }

    let host = match refyard_http::start_http_host(refyard_http::HttpHostOptions {
        service: Arc::clone(&service),
        port: options.port,
        web_root: options.web_root.clone(),
        ticket_ttl_seconds: options.ticket_ttl_seconds,
        grants: Grants::for_repositories(allowed_root_ids.clone(), repository_ids.clone()),
    })
    .await
    {
        Ok(host) => host,
        Err(refyard_http::StartError::PortInUse(port)) if !options.port_explicit => {
            // The default port is a courtesy, not a promise. A second service, a dev
            // server or a tunnel may hold it; taking a free port keeps the default
            // usable, and the note names both numbers so nothing moves silently.
            match refyard_http::start_http_host(refyard_http::HttpHostOptions {
                service: Arc::clone(&service),
                port: 0,
                web_root: options.web_root.clone(),
                ticket_ttl_seconds: options.ticket_ttl_seconds,
                grants: Grants::for_repositories(allowed_root_ids.clone(), repository_ids.clone()),
            })
            .await
            {
                Ok(host) => {
                    stderr(&format!(
                        "note: port {port} is in use; listening on port {} instead (pass --port to pin one)",
                        host.port
                    ));
                    host
                }
                Err(error) => {
                    stderr(&format!("refyard: the listener could not start: {error:?}"));
                    return 2;
                }
            }
        }
        Err(refyard_http::StartError::PortInUse(port)) => {
            stderr(&format!(
                "refyard: port {port} is already in use; stop what is using it or pass a different --port"
            ));
            return 2;
        }
        Err(error) => {
            stderr(&format!("refyard: the listener could not start: {error:?}"));
            return 2;
        }
    };

    let origin = host.base_url.clone();
    // A machine run binds its one ticket to the empty origin: the supervisor spending it
    // sends no Origin header at all, and no browser can spend it, ever.
    let minted = if options.machine {
        host.machine_pairing_url()
    } else {
        host.pairing_url(&origin)
    };
    let pairing_url = match minted {
        Ok(url) => url,
        Err(problem) => {
            stderr(&format!(
                "refyard: a pairing URL could not be minted: {}",
                problem.message
            ));
            return 2;
        }
    };

    // The readiness answer goes out now, while the process lives: a supervisor reading
    // stdout is waiting for exactly this line, and an answer that waits for exit is no
    // answer at all.
    if options.json {
        let readiness = serde_json::json!({
            "serviceInstanceId": host.service_instance_id,
            "port": host.port,
            "url": origin,
            "apiMajor": refyard_host::service::API_MAJOR,
            "ticketTtlSeconds": options.ticket_ttl_seconds,
            "ticketSingleUse": true,
            "repositoryIds": repository_ids,
            "allowedRootIds": allowed_root_ids,
            "ui": options.web_root.as_ref().map(|_| origin.clone()),
            "apiOnly": options.web_root.is_none(),
        });
        stdout(&readiness.to_string());
        // The pairing URL is a secret in transit: in machine mode it goes to stderr, so
        // a supervisor can capture it and a log cannot mistake it for data.
        stderr(&format!("pairing URL (single use): {pairing_url}"));
    } else {
        stdout(&format!(
            "refyard-native serve (api {})\n  repositories: {}\n  port:       {}{}\n  ticket ttl: {}s, single use\n  ui:         {}\n\n  open this URL in your browser to pair this session:\n    {pairing_url}\n\n  ready {}\n  press Ctrl+C to stop\n",
            refyard_host::service::API_MAJOR,
            if options.paths.is_empty() { "(none yet; register through the workbench)".to_string() } else { options.paths.join(", ") },
            host.port,
            if options.port == 0 { " (chosen by the OS)" } else { "" },
            options.ticket_ttl_seconds,
            if options.web_root.is_some() { format!("{origin} (local workbench)") } else { "API-only".to_string() },
            serde_json::json!({
                "serviceInstanceId": host.service_instance_id,
                "port": host.port,
                "url": origin,
            }),
        ));
    }

    if options.open_browser {
        match open_in_browser(&pairing_url) {
            Ok(()) => {}
            Err(reason) => stderr(&format!(
                "note: could not open a browser ({reason}); use the printed URL instead"
            )),
        }
    }

    wait_for_stop().await;
    stderr("stopping: no new requests will be accepted");
    host.close().await;
    0
}

/// Blocks until the process is asked to stop: Ctrl+C on every platform, SIGTERM where
/// the platform has it.
async fn wait_for_stop() {
    #[cfg(unix)]
    {
        use tokio::signal::unix::{signal, SignalKind};
        let mut terminate = signal(SignalKind::terminate()).expect("SIGTERM handler");
        let mut interrupt = signal(SignalKind::interrupt()).expect("SIGINT handler");
        tokio::select! {
            _ = terminate.recv() => {}
            _ = interrupt.recv() => {}
        }
    }
    #[cfg(not(unix))]
    {
        let _ = tokio::signal::ctrl_c().await;
    }
}

/// Opens the pairing URL in this machine's browser: a fixed program per platform, the
/// URL as one argument, no shell anywhere in between.
fn open_in_browser(url: &str) -> Result<(), String> {
    #[cfg(target_os = "macos")]
    let opened = std::process::Command::new("open").arg(url).status();
    #[cfg(all(unix, not(target_os = "macos")))]
    let opened = std::process::Command::new("xdg-open").arg(url).status();
    #[cfg(target_os = "windows")]
    let opened = std::process::Command::new("cmd")
        .args(["/C", "start", "", url])
        .status();
    match opened {
        Ok(status) if status.success() => Ok(()),
        Ok(status) => Err(format!("the opener exited with {status}")),
        Err(error) => Err(error.to_string()),
    }
}

/// The variable that names the SSH configuration file this host should use, shared with
/// the desktop shell's rule: absolute or absent, never relative.
fn ssh_config_from(environment: &[(String, String)]) -> Option<PathBuf> {
    let value = environment
        .iter()
        .find(|(name, _)| name == "REFYARD_SSH_CONFIG")
        .map(|(_, value)| value.as_str())?;
    if value.is_empty() {
        return None;
    }
    let path = PathBuf::from(value);
    if !path.is_absolute() {
        return None;
    }
    Some(path)
}

trait WithSshConfigOpt {
    fn with_ssh_config_opt(self, config: Option<PathBuf>) -> Self;
}

impl WithSshConfigOpt for ApplicationService {
    fn with_ssh_config_opt(self, config: Option<PathBuf>) -> Self {
        match config {
            Some(path) => self.with_ssh_config_file(path),
            None => self,
        }
    }
}
